### PR TITLE
Remark and rehype plugins

### LINKED ISSUE #239: 💡 RFC: provide hook for custom markdown renderer
>  since MD is built-into Astro, it would make sense to provide this hook via some Astro config (or: let you provide a custom Astro plugin that takes over responsibility for rendering MD to HTML)

via [discord](https://discord.com/channels/830184174198718474/845430950191038464/846503698557304832)

### PR DESCRIPTION
**Pull request by akellbl4**
Resolves #239

## Changes

- adds the ability to configure remark plugins
- adds the ability to configure rehype plugins
- adds example with remark plugins

## Testing

- [ ] add tests

## Docs

- [x] add docs

### CODE REVIEW

File: packages/markdown-support/src/types.ts
Comment (natemoo-re) [inline packages/markdown-support/src/types.ts:9]: I think we also want to support `[Promise<{ default: unified.Plugin }>, any]` here so plugins can accept configuration

Comment (natemoo-re) [reply packages/markdown-support/src/types.ts:9]:
```suggestion
  remarkPlugins: (Promise<{ default: unified.Plugin }>|[Promise<{ default: unified.Plugin }>, any])[];
  rehypePlugins: (Promise<{ default: unified.Plugin }>|[Promise<{ default: unified.Plugin }>, any])[];
```

File: packages/markdown-support/src/types.ts
Comment (natemoo-re) [inline packages/markdown-support/src/types.ts:15]: This is purely internal and should be omitted.

Comment (natemoo-re) [reply packages/markdown-support/src/types.ts:15]:
```suggestion
```

Comment (natemoo-re) [reply packages/markdown-support/src/types.ts:15]: Or marked as `@internal`...

```suggestion
  /** @internal */
  $?: {
    scopedClassName: string | null;
  };
```

File: packages/markdown-support/src/index.ts
Comment (natemoo-re) [inline packages/markdown-support/src/index.ts:49]: Is this a breaking change? I still think having GFM and Footnotes out of the box is a good idea, but if you custom `remarkPlugins`, you'd have to include them yourself.

**Review by natemoo-re — CHANGES_REQUESTED** [review]
Overall looks great! A few suggestions, but I think this is really close. Thanks for jumping right on this!

- Could you add a `.astro` page that uses `import { Markdown } from 'astro/components'` to the example here?

**Comment by natemoo-re** [comment]
cc @matthewp if you could take a look at how the remark/rehype plugins are loaded that would be awesome. I feel like this might need to use the `configManager` you introduced?

File: packages/markdown-support/src/types.ts
Comment (akellbl4) [reply packages/markdown-support/src/types.ts:9]: Yep, I had it in mind :) Thanks for the notice!

File: packages/markdown-support/src/index.ts
Comment (akellbl4) [reply packages/markdown-support/src/index.ts:49]: Got it, good idea!

**Comment by akellbl4** [comment]
@natemoo-re thank you for the review, I wanted to make sure if I'm on the right way

File: examples/with-remark-plugins/astro.config.mjs
Comment (matthewp) [inline examples/with-remark-plugins/astro.config.mjs:12]: we should allow this to be a string as well, I think.

File: packages/markdown-support/src/index.ts
Comment (matthewp) [inline packages/markdown-support/src/index.ts:35]: If we allow the config to provide string specifiers this could be: 

```js
const loadedRemarkPlugins = await Promise.all(remarkPlugins.map(plugin => typeof plugin === 'string' ? import(plugin) : plugin));
```

File: packages/markdown-support/src/index.ts
Comment (matthewp) [inline packages/markdown-support/src/index.ts:44]: It seems like remark and rehype plugins are instantiated the same way, in which case is there a reason to differentiate them in the config?

**Comment by matthewp** [comment]
I left some comments.

File: packages/markdown-support/src/index.ts
Comment (akellbl4) [reply packages/markdown-support/src/index.ts:44]: Yes, it works as I said above.

Comment (natemoo-re) [reply packages/markdown-support/src/index.ts:44]: Yep, some plugins require both a `remark` plugin for Markdown and a `rehype` plugin for the HTML, and we need to split them out because of when each one must run.

**Review by matthewp — APPROVED** [review]
lgtm, will defer to @natemoo-re on any changes he thinks are important.

**Review by natemoo-re — COMMENTED** [review]
This looks great! 

Do you want to work on integrating these changes with `import { Markdown } from 'astro/components';` in a `.astro` file or would you be more comfortable if I did that?

**Review by natemoo-re — APPROVED** [review]

**Comment by akellbl4** [comment]
@natemoo-re I would love to help with that.

Should I add docs for the feature?
I'm not really sure how I add tests for it.

**Comment by natemoo-re** [comment]
@akellbl4 Awesome! Why don't you add some docs to this branch and I will add tests? I'm thinking about splitting `<Markdown>` support into another branch.

**Comment by akellbl4** [comment]
Documentation isn't my strong side but I'll give it a try ;)
Okay, I'll provide separate PR for `<Markdown/>`.

**Comment by akellbl4** [comment]
@natemoo-re I've checked `<Markdown/>` and plugins are applied to it as well as to markdown files.

**Comment by natemoo-re** [comment]
@akellbl4 looks amazing—thank you SO much for all your work on this!

I'm going to take care of the tests tomorrow and then this should be good to merge.

**Comment by akellbl4** [comment]
@natemoo-re thank you for your help! It's pleasure to work on the project.
I have a few requests that already exist, for example, this one #397, but it's blocked by another issue. I would love to help with it or whatever blocks it.

**Comment by natemoo-re** [comment]
@akellbl4 Docs look great and I added some tests to make sure there aren't regressions. 

Thanks again for getting this over the line, folks will be really excited about this feature!

**Comment by wooorm** [comment]
Awesome stuff!

**Pull request merged**
